### DISCUSSION TITLE
Trying to extract the stream data directly from I2S stream from audiokit

### DISCUSSION THREAD

**Discussion by Juaspo** [pschatzmann/arduino-audio-tools#1587] [Q&A] [answered]
I am trying to store the data stream from the built in microphone and/or the line in into my own arrays but somehow im unable to do it. I've spent quite some time looking at different examples and managed to get the different streams to work, like streaming to CSV to print on terminal and sending it to I2S output but I'm struggling to get the actual data values as buffer arrays.
I believe that ideally I would create my own class using the class template to manipulate the stream any way I want but that is for now a little bit too advanced for where I am. I have thus far tried to just enable the I2S on the codec chip and trying to use a normal arduino I2S library to access the data directly but this is not giving me the result I'm looking for as I'm only getting 0 as an output despite sending in a sine wave to the board (confirmed sinewave is working).

So far this is the code im testing to try and get this to work. Any help in terms of understanding what I'm doing wrong or what I should try would be most welcome.

```

// Include I2S driver
#include <driver/i2s.h>
#include "AudioBoard.h"

// Connections to A1S
#define I2S_SD 25 //Serial Data or Serial Data in (SDIN)
#define I2S_WS 26 //Word select or left/right clk (LRCLK)
#define I2S_SCK 27 //Serial clock or Bit clock (BCLK)

// Use I2S Processor 0
#define I2S_PORT I2S_NUM_0

// Define input buffer length
#define bufferLen 64
int16_t sBuffer[bufferLen];

/////////////////// start of functions

void i2s_install() {
  // Set up I2S Processor configuration
  const i2s_config_t i2s_config = {
    .mode = i2s_mode_t(I2S_MODE_MASTER | I2S_MODE_RX),
    .sample_rate = 44100,
    .bits_per_sample = i2s_bits_per_sample_t(16),
    .channel_format = I2S_CHANNEL_FMT_ONLY_LEFT,
    .communication_format = i2s_comm_format_t(I2S_COMM_FORMAT_STAND_I2S),
    .intr_alloc_flags = 0,
    .dma_buf_count = 8,
    .dma_buf_len = bufferLen,
    .use_apll = false
  };

  i2s_driver_install(I2S_PORT, &i2s_config, 0, NULL);
}

void i2s_setpin() {
  // Set I2S pin configuration
  const i2s_pin_config_t pin_config = {
    .bck_io_num = I2S_SCK,
    .ws_io_num = I2S_WS,
    .data_out_num = -1,
    .data_in_num = I2S_SD
  };

  i2s_set_pin(I2S_PORT, &pin_config);
}


/////////////////// end of functions

void setup_setup(){
  CodecConfig cfg;
  cfg.input_device = ADC_INPUT_ALL;
  cfg.output_device = DAC_OUTPUT_ALL;
  cfg.i2s.bits = BIT_LENGTH_16BITS;
  cfg.i2s.rate = RATE_44K;
  //cfg.i2s.fmt = I2S_NORMAL;
  //cfg.i2s.mode = MODE_SLAVE; 
  AudioKitEs8388V1.begin(cfg);

  Serial.begin(115200);
  Serial.println("Running I2s to terminal");
  
  // Set up I2S
  i2s_install();
  //i2s_setpin();
  i2s_start(I2S_PORT);
}

void loop_function(){
  //loop function here

  // Get I2S data and place in data buffer
  size_t bytesIn = 0;
  esp_err_t result = i2s_read(I2S_PORT, &sBuffer, bufferLen, &bytesIn, portMAX_DELAY);

  int16_t sXBuffer[bufferLen];
  for (int n=0; n<=bufferLen; n++){
    sXBuffer[n]=sBuffer[n];
  }

  int16_t samples_read = bytesIn;
  if(samples_read > 0) {
    Serial.print("Samples: ");
    Serial.println(samples_read);
    for (int16_t i = 0; i < samples_read; ++i) {
      Serial.println(sXBuffer[i]);
    }
  }
}
```

**Comment by Juaspo** [comment]
Can also add that I was using the code from [this example](https://github.com/pschatzmann/arduino-audio-driver/wiki/Stand%E2%80%90Alone:-Predefined-Boards)

**Comment by pschatzmann** [comment, marked as answer]
Please keep in mind that Audio is using up a lot of RAM so 64 samples will not take you far, so you better use an ESP32 that supports PSRAM. Here is an [example sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-ram-audiokit/streams-audiokit-ram-audiokit.ino)

ps. I dont recommend to use the legacy I2S API: this will not work with the new ESP32 core version any more
